File: Task/Spelling-of-ordinal-numbers/Swift/spelling-of-ordinal-numbers.swift
fileprivate class NumberNames {
    let cardinal: String
    let ordinal: String

    init(cardinal: String, ordinal: String) {
        self.cardinal = cardinal
        self.ordinal = ordinal
    }

    func getName(_ ordinal: Bool) -> String {
        return ordinal ? self.ordinal : self.cardinal
    }

    class func numberName(number: Int, ordinal: Bool) -> String {
        guard number < 100 else {
            return ""
        }
        if number < 20 {
            return smallNames[number].getName(ordinal)
        }
        if number % 10 == 0 {
            return tens[number/10 - 2].getName(ordinal)
        }
        var result = tens[number/10 - 2].getName(false)
        result += "-"
        result += smallNames[number % 10].getName(ordinal)
        return result
    }

    static let smallNames = [
        NumberNames(cardinal: "zero", ordinal: "zeroth"),
        NumberNames(cardinal: "one", ordinal: "first"),
        NumberNames(cardinal: "two", ordinal: "second"),
        NumberNames(cardinal: "three", ordinal: "third"),
        NumberNames(cardinal: "four", ordinal: "fourth"),
        NumberNames(cardinal: "five", ordinal: "fifth"),
        NumberNames(cardinal: "six", ordinal: "sixth"),
        NumberNames(cardinal: "seven", ordinal: "seventh"),
        NumberNames(cardinal: "eight", ordinal: "eighth"),
        NumberNames(cardinal: "nine", ordinal: "ninth"),
        NumberNames(cardinal: "ten", ordinal: "tenth"),
        NumberNames(cardinal: "eleven", ordinal: "eleventh"),
        NumberNames(cardinal: "twelve", ordinal: "twelfth"),
        NumberNames(cardinal: "thirteen", ordinal: "thirteenth"),
        NumberNames(cardinal: "fourteen", ordinal: "fourteenth"),
        NumberNames(cardinal: "fifteen", ordinal: "fifteenth"),
        NumberNames(cardinal: "sixteen", ordinal: "sixteenth"),
        NumberNames(cardinal: "seventeen", ordinal: "seventeenth"),
        NumberNames(cardinal: "eighteen", ordinal: "eighteenth"),
        NumberNames(cardinal: "nineteen", ordinal: "nineteenth")
    ]

    static let tens = [
        NumberNames(cardinal: "twenty", ordinal: "twentieth"),
        NumberNames(cardinal: "thirty", ordinal: "thirtieth"),
        NumberNames(cardinal: "forty", ordinal: "fortieth"),
        NumberNames(cardinal: "fifty", ordinal: "fiftieth"),
        NumberNames(cardinal: "sixty", ordinal: "sixtieth"),
        NumberNames(cardinal: "seventy", ordinal: "seventieth"),
        NumberNames(cardinal: "eighty", ordinal: "eightieth"),
        NumberNames(cardinal: "ninety", ordinal: "ninetieth")
    ]
}

fileprivate class NamedPower {
    let cardinal: String
    let ordinal: String
    let number: UInt64

    init(cardinal: String, ordinal: String, number: UInt64) {
        self.cardinal = cardinal
        self.ordinal = ordinal
        self.number = number
    }

    func getName(_ ordinal: Bool) -> String {
        return ordinal ? self.ordinal : self.cardinal
    }

    class func getNamedPower(_ number: UInt64) -> NamedPower {
        for i in 1..<namedPowers.count {
            if number < namedPowers[i].number {
                return namedPowers[i - 1]
            }
        }
        return namedPowers[namedPowers.count - 1]
    }

    static let namedPowers = [
        NamedPower(cardinal: "hundred", ordinal: "hundredth",
                   number: 100),
        NamedPower(cardinal: "thousand", ordinal: "thousandth",
                   number: 1000),
        NamedPower(cardinal: "million", ordinal: "millionth",
                   number: 1000000),
        NamedPower(cardinal: "billion", ordinal: "billionth",
                   number: 1000000000),
        NamedPower(cardinal: "trillion", ordinal: "trillionth",
                   number: 1000000000000),
        NamedPower(cardinal: "quadrillion", ordinal: "quadrillionth",
                   number: 1000000000000000),
        NamedPower(cardinal: "quintillion", ordinal: "quintillionth",
                   number: 1000000000000000000)
    ]
}

public func numberName(number: UInt64, ordinal: Bool) -> String {
    if number < 100 {
        return NumberNames.numberName(number: Int(truncatingIfNeeded: number),
                                      ordinal: ordinal)
    }
    let p = NamedPower.getNamedPower(number)
    var result = numberName(number: number/p.number, ordinal: false)
    result += " "
    if number % p.number == 0 {
        result += p.getName(ordinal)
    } else {
        result += p.getName(false)
        result += " "
        result += numberName(number: number % p.number, ordinal: ordinal)
    }
    return result
}

func printOrdinal(_ number: UInt64) {
    print("\(number): \(numberName(number: number, ordinal: true))")
}

printOrdinal(1)
printOrdinal(2)
printOrdinal(3)
printOrdinal(4)
printOrdinal(5)
printOrdinal(11)
printOrdinal(15)
printOrdinal(21)
printOrdinal(42)
printOrdinal(65)
printOrdinal(98)
printOrdinal(100)
printOrdinal(101)
printOrdinal(272)
printOrdinal(300)
printOrdinal(750)
printOrdinal(23456)
printOrdinal(7891233)
printOrdinal(8007006005004003)

File: Task/Spelling-of-ordinal-numbers/C-sharp/spelling-of-ordinal-numbers.cs
using System;
using System.Collections.Generic;

class SpellingOfOrdinalNumbers
{
    private static readonly Dictionary<string, string> ordinalMap = new Dictionary<string, string>
    {
        {"one", "first"},
        {"two", "second"},
        {"three", "third"},
        {"five", "fifth"},
        {"eight", "eighth"},
        {"nine", "ninth"},
        {"twelve", "twelfth"}
    };

    static void Main(string[] args)
    {
        long[] tests = new long[] { 1, 2, 3, 4, 5, 11, 65, 100, 101, 272, 23456, 8007006005004003L };
        foreach (long test in tests)
        {
            Console.WriteLine($"{test} = {ToOrdinal(test)}");
        }
    }

    private static string ToOrdinal(long n)
    {
        string spelling = NumToString(n);
        string[] split = spelling.Split(' ');
        string last = split[split.Length - 1];
        string replace;
        if (last.Contains("-"))
        {
            string[] lastSplit = last.Split('-');
            string lastWithDash = lastSplit[1];
            string lastReplace;
            if (ordinalMap.ContainsKey(lastWithDash))
            {
                lastReplace = ordinalMap[lastWithDash];
            }
            else if (lastWithDash.EndsWith("y"))
            {
                lastReplace = lastWithDash.Substring(0, lastWithDash.Length - 1) + "ieth";
            }
            else
            {
                lastReplace = lastWithDash + "th";
            }
            replace = lastSplit[0] + "-" + lastReplace;
        }
        else
        {
            if (ordinalMap.ContainsKey(last))
            {
                replace = ordinalMap[last];
            }
            else if (last.EndsWith("y"))
            {
                replace = last.Substring(0, last.Length - 1) + "ieth";
            }
            else
            {
                replace = last + "th";
            }
        }
        split[split.Length - 1] = replace;
        return string.Join(" ", split);
    }

    private static readonly string[] nums = new string[]
    {
        "zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine",
        "ten", "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen"
    };

    private static readonly string[] tens = new string[] { "zero", "ten", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety" };

    private static string NumToString(long n)
    {
        return NumToStringHelper(n);
    }

    private static string NumToStringHelper(long n)
    {
        if (n < 0)
        {
            return "negative " + NumToStringHelper(-n);
        }
        if (n <= 19)
        {
            return nums[n];
        }
        if (n <= 99)
        {
            return tens[n / 10] + (n % 10 > 0 ? "-" + NumToStringHelper(n % 10) : "");
        }
        string label = null;
        long factor = 0;
        if (n <= 999)
        {
            label = "hundred";
            factor = 100;
        }
        else if (n <= 999999)
        {
            label = "thousand";
            factor = 1000;
        }
        else if (n <= 999999999)
        {
            label = "million";
            factor = 1000000;
        }
        else if (n <= 999999999999L)
        {
            label = "billion";
            factor = 1000000000;
        }
        else if (n <= 999999999999999L)
        {
            label = "trillion";
            factor = 1000000000000L;
        }
        else if (n <= 999999999999999999L)
        {
            label = "quadrillion";
            factor = 1000000000000000L;
        }
        else
        {
            label = "quintillion";
            factor = 1000000000000000000L;
        }
        return NumToStringHelper(n / factor) + " " + label + (n % factor > 0 ? " " + NumToStringHelper(n % factor) : "");
    }
}
